### PR TITLE
Default to only applying QK norm over the head dimension

### PR DESCRIPTION
**Pull request by christianazinn**
Before this change, the default application of QK norm was to apply it to dimensions `[*, N, H]`. This caused problems with GQA, in which the K tensor is of shape `[T, B, N_kv, H]`, which doesn't match. This PR fixes that by changing QK norm to normalize over `[*, H]` by default, as in e.g. Gemma3 and the original QK norm paper, with the option to go back to using `[*, N, H]` available as `qk_layernorm_over_heads`.

### CODE REVIEW

**Comment by StellaAthena** [comment]
@christianazinn why is this a better default than the current defualt?

**Comment by christianazinn** [comment]
> @christianazinn why is this a better default than the current defualt?

I had assumed that most models used this normalization style, but it seems that's not the case (e.g. Olmo2). Perhaps it ought to be changed to a toggleable option that is enforced for GQA.

**Comment by aflah02** [comment]
+1
I ran a deep research query and attached below is the overall picture across frameworks and models (I verified the outputs) - 

Model Response:

**Short Answer**

QK normalization is usually applied **independently per token** to the query and key activations. The key implementation difference is the feature axis:

| Model/framework | Q/K tensor at norm time | Normalized axis | Mixes heads in norm? |
|---|---:|---:|---:|
| Original QKNorm paper | per-head Q/K vectors | `head_dim` | No |
| ViT-22B / Dehghani-style QK LayerNorm | Q/K projection vectors, conceptually per query/key dimension | usually `head_dim` in multi-head implementations | No in typical implementations |
| Qwen3 / Marin 32B | `[B, T, H, head_dim]` | `head_dim` | No |
| TorchTitan | `[B, T, H, head_dim]` | `head_dim` | No |
| Megatron-Core built-in `qk_layernorm` | `[S, B, H, head_dim]` | `head_dim` | No |
| OLMo-core default / HF OLMo2 | `[B, T, H * head_dim]` before reshape | all Q heads flattened for Q; all KV heads flattened for K | Yes |
| MiniMax-M2 Megatron Bridge special case | full Q/K projection | `num_heads * head_dim` | Yes |

**Mental Model**

If your Q tensor is shaped:

```text
q: [batch, sequence, num_q_heads, head_dim]
k: [batch, sequence, num_kv_heads, head_dim]
```

then “standard modern QK norm” usually means:

```python
q = rmsnorm(q)  # normalized over last dim: head_dim
k = rmsnorm(k)  # normalized over last dim: head_dim
```

That is **per token, per head**, over the vector that participates in one dot product. Batch, sequence, and head index are not reduced.

But some models, especially **OLMo2-style**, do:

```python
q = rmsnorm(q_flat)  # q_flat: [batch, sequence, num_q_heads * head_dim]
k = rmsnorm(k_flat)  # k_flat: [batch, sequence, num_kv_heads * head_dim]
q = q.view(batch, sequence, num_q_heads, head_dim)
k = k.view(batch, sequence, num_kv_heads, head_dim)
```

That **does couple heads through the RMS/LN denominator** before attention.

**Original Papers**

The 2020 “Query-Key Normalization for Transformers” paper is unambiguous: it applies **L2 normalization along the head dimension** of each query and key matrix, then uses a learnable scale instead of the usual fixed `1 / sqrt(d_k)` factor. So this is per-head vector normalization, not flatten-all-heads normalization. ([arxiv.org](https://arxiv.org/abs/2010.04245))

The ViT-22B paper popularized the “QK LayerNorm” phrase in large-scale transformer training. It says LayerNorm is applied to queries and keys before dot-product attention and writes the attention as `softmax(1/sqrt(d) * LN(XWQ) LN(XWK)^T)`, where `d` is the query/key dimension. In multi-head attention, that maps naturally to the per-head query/key vector dimension, although some later LLM codebases chose full-projection variants. ([sciencestack.ai](https://www.sciencestack.ai/paper/2302.05442v1?utm_source=openai))

**OLMo**

OLMo-core supports both variants, but its default is important. In `Attention.__init__`, if `use_head_qk_norm=True`, it builds Q/K norms with size `head_dim`; otherwise it builds Q norm with `n_heads * head_dim` and K norm with `n_kv_heads * head_dim`. ([olmo-core.readthedocs.io](https://olmo-core.readthedocs.io/en/latest/_modules/olmo_core/nn/attention.html))

In the forward path, OLMo-core applies the default full-dimension Q/K norm **before** reshaping Q and K into heads. Only when `use_head_qk_norm=True` does it reshape to `[B, T, H, head_dim]` first and then apply the norm per head. ([olmo-core.readthedocs.io](https://olmo-core.readthedocs.io/en/latest/_modules/olmo_core/nn/attention.html))

Hugging Face’s OLMo2 docs say OLMo2 adds norm on attention queries and keys, and the modular Transformers example shows `q_norm = Olmo2RMSNorm(config.num_attention_heads * self.head_dim)`, which is the full flattened Q projection dimension. ([huggingface.co](https://huggingface.co/docs/transformers/v4.48.1/model_doc/olmo2?utm_source=openai))

So for **OLMo2/OLMo-core default**: QK norm is **over all Q heads together for Q**, and **over all KV heads together for K**, per token. It is not per attention head unless `use_head_qk_norm=True`.

**Marin**

Marin 32B’s model card describes the architecture as **Qwen3-style 32B with QK-Norm attention**, with 40 attention heads and 8 KV heads. ([huggingface.co](https://huggingface.co/marin-community/marin-32b-base))

Qwen3’s Transformers implementation creates `q_norm` and `k_norm` with `head_dim`, then applies them after reshaping the projections to `(..., -1, head_dim)` and before transposing to attention layout. That means Marin’s Qwen3-style QK norm is **per token, per Q/K head, over `head_dim`**, not across flattened heads. ([github.com](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3/modeling_qwen3.py?utm_source=openai))

**Megatron-Core**

Megatron-Core’s public config has `qk_layernorm`, described as applying the selected normalization type to query and key embeddings, plus `qk_l2_norm` for Llama-4-style L2 QK norm. ([docs.nvidia.com](https://docs.nvidia.com/megatron-core/developer-guide/nightly/apidocs/core/core.transformer.transformer_config.html?utm_source=openai))

The actual built-in self-attention constructs Q/K norm with `hidden_size=self.hidden_size_per_attention_head`, then reshapes Q to `[seq, batch, heads, head_dim]` and applies `q_layernorm` and `k_layernorm` there. That makes Megatron’s built-in path **per head over `head_dim`**. ([raw.githubusercontent.com](https://raw.githubusercontent.com/NVIDIA/Megatron-LM/main/megatron/core/transformer/attention.py))

NVIDIA’s Megatron Bridge documentation explicitly calls this out: Megatron’s built-in QK norm is per-head with weight shape `head_dim`, whereas the MiniMax-M2 bridge implements full-dimension RMSNorm over `num_heads * head_dim` before splitting into heads. ([docs.nvidia.com](https://docs.nvidia.com/nemo/megatron-bridge/0.4.0/apidocs/bridge/bridge.models.minimax_m2.minimax_m2_provider.html))

**TorchTitan**

TorchTitan’s common attention path projects Q/K/V, reshapes each to `[B, L, local_heads, head_dim]`, and returns those shaped tensors from the QKV projection layer. ([raw.githubusercontent.com](https://raw.githubusercontent.com/pytorch/torchtitan/main/torchtitan/models/common/attention.py))

Its `GQAttention` then applies optional QK normalization to `xq` and `xk` before RoPE. Since `xq` and `xk` are already shaped with `head_dim` last, and TorchTitan uses `RMSNorm`, this is **per token, per head, over the last dimension (`head_dim`)**. ([raw.githubusercontent.com](https://raw.githubusercontent.com/pytorch/torchtitan/main/torchtitan/models/common/attention.py))

**Bottom Line**

When someone says “QK norm” in LLM code, do not assume one axis. Check the norm module’s `normalized_shape` and whether it is applied before or after `view(..., num_heads, head_dim)`.

The two common patterns are:

```python
# Per-head QK norm: Qwen3, Marin, TorchTitan, Megatron built-in
q = q_proj(x).view(B, T, Hq, Dh)
k = k_proj(x).view(B, T, Hkv, Dh)
q = q_norm(q)  # normalized over Dh
k = k_norm(k)  # normalized over Dh
```

```python
# Full-projection QK norm: OLMo2-style, MiniMax-M2 special handling
q = q_norm(q_proj(x))  # normalized over Hq * Dh
k = k_norm(k_proj(x))  # normalized over Hkv * Dh
q = q.view(B, T, Hq, Dh)
k = k.view(B, T, Hkv, Dh)
```

So your intuition is right: in some places it is **per attention head**, and in others it is **after flattening all heads for Q or K**. For the examples you listed: **Megatron and TorchTitan are per-head; Marin/Qwen3 is per-head; OLMo2-style is flattened-across-heads by default.**

**Comment by aflah02** [comment]
My proposal:

- Have toggle b/w both choices (as an enum incase there are more variants added later)
- Have an arg for separate norms. This allows user to have a separate q norm and a separate k norm
- If user chooses the current default with GQA then construct separate norms instead of erroring out similar to OLMo but document this behaviour clearly. Here's how OLMo does it for reference -

```
        self.q_norm: Optional[LayerNorm] = None
        self.k_norm: Optional[LayerNorm] = None
        if qk_norm is not None:
            if use_head_qk_norm:
                self.q_norm = qk_norm.build(size=self.head_dim, init_device=init_device)
                self.k_norm = qk_norm.build(size=self.head_dim, init_device=init_device)
            else:
                self.q_norm = qk_norm.build(size=n_heads * self.head_dim, init_device=init_device)
                self.k_norm = qk_norm.build(
                    size=self.n_kv_heads * self.head_dim, init_device=init_device
                )
```


Src: https://olmo-core.readthedocs.io/en/latest/_modules/olmo_core/nn/attention.html

**Comment by christianazinn** [comment]
@aflah02 tried implementing w.r.t. your proposal, please lmk

**Comment by aflah02** [comment]
Thanks @christianazinn!
Added a comment otherwise it looks good to me. Ofc would have to test to make sure everything runs fine but this is what I had in mind.

**Comment by aflah02** [comment]
Hi @christianazinn 
Are you interested in taking this forward? If not then no worries and I can do the finishing touches as this is relevant for another PR - https://github.com/EleutherAI/gpt-neox/pull/1408

**Comment by christianazinn** [comment]
@aflah02 If you could take it from here that'd be great! Sorry for my inactivity on this.

**Comment by aflah02** [comment]
> @aflah02 If you could take it from here that'd be great! Sorry for my inactivity on this.

No worries! Thanks for the help

**Comment by aflah02** [comment]
Closing this and continuing work at https://github.com/EleutherAI/gpt-neox/pull/1424
